arch/common/ramfunc.ld

```diff
@@ -7,9 +7,23 @@
 
 /* Copied from linker.ld */
 
-SECTION_DATA_PROLOGUE(.ramfunc,,)
+/* The ramfunc region must have an associated MPU entry added for it, so that
+ * code can be executed in RAM. This means the start of the ramfunc region
+ * must respect MPU alignment requirements. We include a padding section here
+ * in order to make sure the region will be placed at the correct offset
+ * in flash.
+ */
+SECTION_PROLOGUE(ramfunc_padding_section,,)
 {
 	MPU_ALIGN(__ramfunc_size);
+#ifdef CONFIG_BUILD_ALIGN_LMA
+} GROUP_DATA_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
+#else
+} GROUP_DATA_LINK_IN(RAMABLE_REGION, RAMABLE_REGION)
+#endif
+
+SECTION_DATA_PROLOGUE(.ramfunc,,)
```

Comment:
Would it be better to align using the header? e.g. `SECTION_DATA_PROLOGUE(.ramfunc,,MPU_ALIGN_HDR(__ramfunc_size))` where you need to define `MPU_ALIGN_HDR(sz)` to be `ALIGN(sz)` (instead of `. = ALIGN(sz)`). Using the header avoids an additional section.

Reply:
Using this, we will run into an issue with `SECTION_DATA_PROLOGUE`. On  Cortex-M, this expands to `name options: ALIGN_WITH_INPUT align`. So if we used something like `SECTION_DATA_PROLOGUE(.ramfunc,,ALIGN(32)`, we would get the following: `.ramfunc: ALIGN_WITH_INPUT ALIGN(32)`. According to the documentation for GNU LD, we cannot define both `ALIGN_WITH_INPUT` and `ALIGN(x)` for a given section: https://sourceware.org/binutils/docs/ld/Output-Section-Attributes.html. It does seem we could use `SUBALIGN(sz)` here, but this will force each input section to be aligned to a boundary of `sz`, which is a more strict restriction than we need, and would result in increased ROM utilization